📚 Documentation: Tutorial for Next JS

Mridul2820 opened the pull request:
## What does this PR do?

Added Tutorial for a functioning app with Next JS in the Docs

## Test Plan

The Tutorial can be found in the route `/docs/tutorials/nextjs/step-1`

## Related PRs and Issues

#80 

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-1/+page.markdoc:
```diff
@@ -0,0 +1,28 @@
+description: Build a Next JS project using Appwrite.
+step: 1
+difficulty: beginner
+readtime: 10
```
I would probably say this is not a 10 minute read

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-5/+page.markdoc:
```diff
@@ -0,0 +1,220 @@
+Now that we have created a collection, we have to add permissions to the collection. Go to the settings tab, scoll to the permissions sections add the following permissions:
+
+{% only_dark %}
+![Create project screen](/images/docs/tutorials/dark/collection-permissions.png)
```
Please provide the original taken from a 1400x900 viewport with chrome/firefox devtools.

I will redo these with our branding assets.

Mridul2820 replied:
Added all the screenshots in 1400x900 viewport

gewenyu99 commented on static/images/docs/tutorials/expense-tracker.png:
Can I also have the original here so I can redo the screenshot to be consistent with rest of tutorials

1400 x 900 viewport, thank you

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-4/+page.markdoc:
```diff
@@ -0,0 +1,206 @@
+---
+layout: tutorial
+title: Implement Authtentication
```
Titles should be in sentence case please, thank you!

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-4/+page.markdoc:
```diff
@@ -0,0 +1,206 @@
+
+# Signup Page {% #signup-page %}      
+
+In your project under the pages directory, create a new file `src/pages/signup.js`. Next JS will automatically create a route for this page `/signup`. Let's create the signup form.
```
Should be `Next.js`, please check all occurrences.

gewenyu99 reviewed (changes requested):
Another issue I see is that this is using Tailwind. We prefer very little styling or to use https://pink.appwrite.io/.

We prefer to not recommend style library in docs, which is out of scope.

Also pay attention to Titles and headings, should be **sentence case**

Mridul2820 commented:
Hi @gewenyu99 , I have added all the changes you asked for

- Added all the images in 1400x900 viewport
- Titles are now in sentence case
- Changed Next JS to Next.js
- Used https://pink.appwrite.io/. for styling and styles are minimal with the Appwrite theme
- Increased the Read Time
- Changed user to Walter O'Brien for screenshots

Here is the current preview of the project

<img width="700" alt="expense-tracker" src="https://github.com/appwrite/website/assets/43516008/6fe275c0-7320-4787-8b21-bad0034026e5">

gewenyu99 commented:
> Hi @gewenyu99 , I have added all the changes you asked for
> 
> * Added all the images in 1400x900 viewport
> * Titles are now in sentence case
> * Changed Next JS to Next.js
> * Used https://pink.appwrite.io/. for styling and styles are minimal with the Appwrite theme
> * Increased the Read Time
> * Changed user to Walter O'Brien for screenshots
> 
> Here is the current preview of the project
> 
> <img alt="expense-tracker" width="700" src="https://user-images.githubusercontent.com/43516008/272489166-6fe275c0-7320-4787-8b21-bad0034026e5.png">

Hi I meant these images should be uncropped. We have a tool that crops them or frames them consistently, can I have the originals please? Thanks! What ever the output of the screenshot tool from your browser gives you! 

Thanks ❤️ 

https://github.com/appwrite/website/assets/29069505/8a712dae-cbb0-4ccc-a843-e4090a5f7b3a

Mridul2820 commented:
Hi @gewenyu99  I have added the raw screenshot images the way you showed. Let me know if this works.

gewenyu99 commented:
@Mridul2820 Thank you so much.

Which browsers and operating system do you use? The links in the top nav of the console is blue for some reason, this appears to be a bug on our Console that we might need to fix. 

I will retake these screenshots for you sometime later, because it doesn't look right on your screen (which is our fault :P)

Thanks!

Mridul2820 commented:
I have used Mac M1 and chrome browser to take the screenshots

Let me know how we can fix the screenshot issue :)

gewenyu99 commented:
> I have used Mac M1 and chrome browser to take the screenshots
> 
> Let me know how we can fix the screenshot issue :)

That's really weird. The text color from the links are wrong, it should be white, not pink. No idea why they're wrong. I might have to take these screenshots myself, so it's out of your hand.

When I get some time I will be reviewing the writing in more detail, thanks!

ArmanNik commented:
Hey @Mridul2820 thank you for your contribution :raised_hands: 

I'm trying to figure out why the styles are not loading correctly for you. Do you get the same issue if you open the console on an incognito page?

Mridul2820 commented:
Hey @ArmanNik I dig into that issue and found that it's from my side. There is some chrome extension that was causing the issue. I have taken the screenshots again and reuploaded them:)

Please check them and let me know if these screenshots are fine.

gewenyu99 commented:
> Hey @ArmanNik I dig into that issue and found that it's from my side. There is some chrome extension that was causing the issue. I have taken the screenshots again and reuploaded them:)
> 
> Please check them and let me know if these screenshots are fine.

Let me take a look!

gewenyu99 commented:
Perfect, screenshots look fine. Looks great so far, I'll find time to follow this and make sure the experience is good soon. Thanks a lot @Mridul2820

gewenyu99 commented:
Okay fantastic, One thing I'm gonna ask you to do is clean this folder up a little
<img width="326" alt="Screenshot 2023-10-06 at 5 14 07 PM" src="https://github.com/appwrite/website/assets/29069505/f5ccd147-697a-4367-94f3-d32064fd60b3">

gewenyu99 commented:
There seems to be not the same number of light and dark mode images
Which is weird.

gewenyu99 commented:
![brandbird (36)](https://github.com/appwrite/website/assets/29069505/b927824d-330d-4411-b60e-c221664377e5)
![brandbird (35)](https://github.com/appwrite/website/assets/29069505/a0bcf21f-e560-421b-853c-657ef6221feb)

gewenyu99 commented:
![brandbird (38)](https://github.com/appwrite/website/assets/29069505/c87e17b2-e22b-49c2-b75b-66f60726d9f0)
![brandbird (37)](https://github.com/appwrite/website/assets/29069505/c6f9351f-6f3f-4f9b-96d4-8934697af081)

Here are your formatted screenshots

Can I also have a full screenshot of your app? It looks super cool, would love to show it off on the first page of your tutorial!

Mridul2820 commented:
I have added the formatted screenshots and Also added my app screenshot on the first page.

Could you give me a small suggestion, Should I provide the full code of the project? If yes, How can I provide the link to the full code? I have the full project code on my local computer.

gewenyu99 commented:
@Mridul2820 I would create a repo, we can fork it later from the Appwrite org :)

But you don't need to provide the full code in your tutorial.

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-2/+page.markdoc:
````diff
@@ -0,0 +1,70 @@
+Create a new Next.js project using the `create-next-app` command.
+
+```sh
+npx create-next-app expense-tracker --eslint
````
What are the options you plan to take?

for example ```
✔ Would you like to use TypeScript? … No / Yes
✔ Would you like to use Tailwind CSS? … No / Yes
✔ Would you like to use `src/` directory? … No / Yes
✔ Would you like to use App Router? (recommended) … No / Yes
✔ Would you like to customize the default import alias (@/*)? … No / Yes
Creating a new Next.js app in /Users/wen/Documents/GitHub/expense-tracker.
```

smthing like this might be helpful

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-2/+page.markdoc:
````diff
@@ -0,0 +1,70 @@
+We will be using cookie to store the session token. So we need to install `next-cookie` package. We will also be using `react-hot-toast` to show toast messages.
+
+```bash
+npm install next-cookie react-hot-toast
````
Do we need to do this? I thought your browser would handle cookies for you, unless you're using SSR?

Mridul2820 replied:
Yes, but in Next.js we are storing some details of user that why we need this

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-2/+page.markdoc:
````diff
@@ -0,0 +1,70 @@
+npm install next-cookie react-hot-toast
+```
+
+We also have to add the `Toaster` component in the `pages/_app.js` file. This component will render all toasts.
````
I don't have a pages directory that gets generated, where does this file come from?

Mridul2820 replied:
In the app creation time, we must check the `pages` directory to yes. I am using the Next.js `pages` router instead of the `app` router as it is beginner's friendly.

I can change the tutorial to the `app` directory if you want.

gewenyu99 reviewed (changes requested):
Hi I tried following your tutorial but I ran into a few issues. Let's take a look at these first and I'll re review, thanks!

Mridul2820 commented:
> <img alt="Screenshot 2023-10-10 at 5 05 06 PM" width="809" src="https://user-images.githubusercontent.com/29069505/274064134-122365d8-b8d8-40a7-a77b-a377917d6d49.png"> Can I get an original for this screenshot? Smthing's not right, let me edit it again.

![localhost_3000_](https://github.com/appwrite/website/assets/43516008/5ef6d708-78b4-47ef-9ebf-27506f9a5ce3)

Mridul2820 commented on static/images/docs/tutorials/expense-tracker-collection.png:
This image has already deleted

gewenyu99 commented:
@Mridul2820 Where did the `pages/_app.js` directory and file come from? My generated project doesn't have these, and creating them will not run.
<img width="265" alt="Screenshot 2023-10-11 at 3 14 26 PM" src="https://github.com/appwrite/website/assets/29069505/b3e27f29-8efb-4d77-87bb-b0ee38dc10c2">

Mridul2820 commented:
@gewenyu99 In the Next.js project creation time you have set this step to 
```
✔ Would you like to use App Router? (recommended) … No / Yes: NO
```

 I am using the Next.js `pages` router instead of the `app` router as it is beginner's friendly. But I will convert it to an `app` directory tutorial as the Appwite quickstart also has the same

gewenyu99 commented:
> @gewenyu99 In the Next.js project creation time you have set this step to
> 
> ```
> ✔ Would you like to use App Router? (recommended) … No / Yes: NO
> ```
> 
> I am using the Next.js `pages` router instead of the `app` router as it is beginner's friendly. But I will convert it to an `app` directory tutorial as the Appwite quickstart also has the same

Ahhh this is what I'm missing. 😅 hahaha I feel dumb! Yeah if you could make this a natural progression from the quick start, it would be best. I think at the end of the day, someone will follow this to learn good practices, right? I assumed Next.js's router is one of it's main attractions over react.

Mridul2820 commented:
Sure, there are some minor changes I will do it ASAP

Mridul2820 commented:
@gewenyu99 I have updated the tutorial with Next.js `app` router.

gewenyu99 commented:
Great @Mridul2820 This is looking really good from my side, before we go into the super details for the writing, let me get a review from a frontend engineer ;)

The big items have been addressed. Good job!

TGlide commented on src/routes/docs/tutorials/nextjs/step-4/+page.markdoc:
```diff
@@ -0,0 +1,219 @@
+import { account } from "@/config/appwrite";
+
+const SignUp = (props) => {
+  const [name, setName] = useState("");
+  const [email, setEmail] = useState("");
+  const [password, setPassword] = useState("");
```
We don't need to use state variables for forms in this case, you can use the [formData on the submit function](https://react.dev/reference/react-dom/components/input#reading-the-input-values-when-submitting-a-form)

TGlide commented on src/routes/docs/tutorials/nextjs/step-4/+page.markdoc:
```diff
@@ -0,0 +1,219 @@
+                placeholder="Name"
+                id="name"
+                type="text"
+                required={true}
```
I think you can just do

```suggestion
                required
```

TGlide commented on src/routes/docs/tutorials/nextjs/step-4/+page.markdoc:
````diff
@@ -0,0 +1,219 @@
+};
+```
+
+After the user is created, we are creating a session using `createEmailSession` method. Then we are creating a JWT token using `createJWT` method. We are storing the JWT token and user id in the cookie. Then we are redirecting the user to the home page.
````
I think this can be better explained, e.g. why are we doing these things?

```suggestion
First, we need to create the user in our Appwrite instance. After the user is created, we create a session using `createEmailSession` method, to effectively login the user into our app. We then create a JWT token using the `createJWT` method, and store the token and user ID into a cookie. This allows the user to remain logged in, even if the app is closed.

With all that done, we can redirect the user to the home page.
```

TGlide commented on src/routes/docs/tutorials/nextjs/step-4/+page.markdoc:
```diff
@@ -0,0 +1,219 @@
+import { account } from "@/config/appwrite";
+
+const Login = (props) => {
+  const [email, setEmail] = useState("");
+  const [password, setPassword] = useState("");
```
Same here, no need for state

TGlide commented on src/routes/docs/tutorials/nextjs/step-4/+page.markdoc:
```diff
@@ -0,0 +1,219 @@
+
+import { account } from "@/config/appwrite";
+
+const Login = (props) => {
```
This makes me think that we should have *some* form of typing, even if it's using JSDoc rather than TS for full compatibility.

TGlide commented on src/routes/docs/tutorials/nextjs/step-4/+page.markdoc:
````diff
@@ -0,0 +1,219 @@
+export default Login;
+```
+
+Now let's add the login function. We will be using Appwrite's `createEmailSession` method to create a new session. We will also be using `react-hot-toast` to show toast messages.
````
No need to specify `react-hot-toast` again I'd say

TGlide commented on src/routes/docs/tutorials/nextjs/step-5/+page.markdoc:
````diff
@@ -0,0 +1,278 @@
+Now you will get a database ID and collection ID. Copy the these two and save it in the `/src/config/appwrite.js` file.
+
+```js
+export const EXPENSE_DATABASE_ID = "<YOUR_DATABASE_ID>";
+export const EXPENSE_COLLECTION_ID = "<YOUR_COLLECTION_ID>";
````
Again, we should probably use env vars

TGlide commented on src/routes/docs/tutorials/nextjs/step-5/+page.markdoc:
```diff
@@ -0,0 +1,278 @@
+  account.client.setJWT(jwt);
+
+  const body = await req.json();
+  const { userId, title, amount, type } = body;
+  if (!userId || !title || !amount || !type) {
+    return new NextResponse("Missing fields", { status: 401 });
+  }
```
These should probably be checked *before* setting the JWT

TGlide replied:
Also, add a space between these blocks :)
```suggestion
  const { userId, title, amount, type } = body;
  if (!userId || !title || !amount || !type) {
    return new NextResponse("Missing fields", { status: 401 });
  }
  
```

TGlide commented on src/routes/docs/tutorials/nextjs/step-5/+page.markdoc:
```diff
@@ -0,0 +1,278 @@
+      EXPENSE_DATABASE_ID,
+      EXPENSE_COLLECTION_ID,
+      ID.unique(),
+      {
+        userId: userId,
+        title: title,
+        amount: amount,
+        type: type,
+      }
```
```suggestion
      { userId, title, amount, type }
```

TGlide commented on src/routes/docs/tutorials/nextjs/step-5/+page.markdoc:
```diff
@@ -0,0 +1,278 @@
+    );
+    return new NextResponse("Saved successfully", { status: 200 });
+  } catch (error) {
+    console.log(error);
```
Is this leftover?

TGlide commented on src/routes/docs/tutorials/nextjs/step-5/+page.markdoc:
````diff
@@ -0,0 +1,278 @@
+}
+```
+
+We have used Appwrite `createDocument` method to add a new document to the database. The method takes the following parameters: `databaseId`, `collectionId`, `documentId`, `data`. 
+
+The `databaseId` is the ID of the database we want to add the document to. The `collectionId` is the ID of the collection we want to add the document to. The `documentId` is the ID of the document we want to add. We have used the `ID.unique()` method to generate a unique ID for the document. The `data` is an object that contains the data we want to add to the document.
+
+We have created a new API endpoint that will add a new expense to the database. The endpoint will be available at `/api/expense`.
````
```suggestion
In this endpoint, we use the `createDocument` method to add a new document to the database, which accepts the following parameters: `databaseId`, `collectionId`, `documentId`, `data`. 

In Appwrite, documents live in a collection, which in turn live in a database. So to add a document, we need its ID, as well as the ID of its containing collection, and the ID of the database that contains said collection. We also need the data that the document contains.

Since we're creating a new document, it doesn't yet have an ID. We can use the `ID.unique()` method to generate a unique one for it.

The document created by this endpoint represents a new expense. The endpoint is now available at `http://localhost:3000/api/expense`.
```

TGlide commented on src/routes/docs/tutorials/nextjs/step-5/+page.markdoc:
```diff
@@ -0,0 +1,278 @@
+  const handleSubmit = async (e) => {
+    e.preventDefault();
+
+    if (!item.title) {
+      return toast.error("Title is required");
+    }
+    if (!item.amount) {
+      return toast.error("Amount is required");
+    }
+    if (!item.type) {
+      return toast.error("Type is required");
+    }
```
This should probably be a concern of `createExpense`. In the case there's an error there, you can catch the error here, and use a toast. Especially because other errors may heppen, e.g. a network failure

TGlide commented on src/routes/docs/tutorials/nextjs/step-5/+page.markdoc:
```diff
@@ -0,0 +1,278 @@
+  const jwt = cookie.get("jwt");
+  const userId = cookie.get("userid");
+
+  const createExpense = async (title, amount, type) => {
```
Maybe you can accept a single item argument?

TGlide commented on src/routes/docs/tutorials/nextjs/step-5/+page.markdoc:
```diff
@@ -0,0 +1,278 @@
+import toast from "react-hot-toast";
+
+const CreateExpensePopup = ({ setPopup, createExpense }) => {
+  const [item, setItem] = useState({
+    title: "",
+    amount: "",
+    type: "",
+  });
```
No need for state

TGlide commented on src/routes/docs/tutorials/nextjs/step-5/+page.markdoc:
```diff
@@ -0,0 +1,278 @@
+
+We are using the `createExpense` function on the `handleSubmit` function to create a new expense.
+
+Here we are also using a custom className `lite-bg` which we can define in the `/src/styles/globals.css` file.
```
We should probably setup globals.css at the very beginning of the tutorial

TGlide commented on src/routes/docs/tutorials/nextjs/step-5/+page.markdoc:
```diff
@@ -0,0 +1,278 @@
+---
+layout: tutorial
+title: Add database
```
Maybe we can call this adding expenses? Since the overall is that we create a database so that we can create our expenses

TGlide commented on src/routes/docs/tutorials/nextjs/step-5/+page.markdoc:
````diff
@@ -0,0 +1,278 @@
+}
+```
+
+Now your expense createtion form is created, you can add a new expense to the database.
````
Maybe we can also alert to the fact that when you do create an expense, you should be able to see it in the AW dashboard?

TGlide reviewed (commented):
Hi, thank you so much for the PR! Amazing work 🎉 

I've left some comments. I've reviewed most pages, and will review the final ones once the comments have been all addressed.

Would you by any chance have a repository with the final code available? @gewenyu99 maybe we should put that into a Appwrite repo?

Mridul2820 commented:
> Hi, thank you so much for the PR! Amazing work 🎉
> 
> I've left some comments. I've reviewed most pages, and will review the final ones once the comments have been all addressed.
> 
> Would you by any chance have a repository with the final code available? @gewenyu99 maybe we should put that into a Appwrite repo?

Yes, I have created a repository for this tutorial [expense-tracker](https://github.com/Mridul2820/expense-tracker). I would be happy if you guys put it in an Appwrite Repo.
Also, I will resolve the comments as well

Mridul2820 commented:
@TGlide I have added the changes as per your comments

gewenyu99 commented:
@Mridul2820 I'm going to take another review. Thank you so much for your hard work, I know this is a lot of back and forth, but we hold docs PRs to a very high standards because it influences every Appwrite user following your contribution.

Thanks again!

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-4/+page.markdoc:
```diff
@@ -0,0 +1,218 @@
+step: 4
+---
+
+# Signup page {% #signup-page %}      
```
<img width="860" alt="Screenshot 2023-10-13 at 3 59 55 PM" src="https://github.com/appwrite/website/assets/29069505/234c1233-1490-449c-8ea6-0c06e7d017ff">

Mridul2820 replied:
Can you give me more details on this? I have added cookie import both login and signup component

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-7/+page.markdoc:
```diff
@@ -0,0 +1,410 @@
+  database,
+} from "@/config/appwrite";
+
+// Post - create expense
```
Think about how to make this more visible, 

maybe 

```
// Post - create expense
export async function POST(req) {
  ...
  }
  ```

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-7/+page.markdoc:
```diff
@@ -0,0 +1,410 @@
+
+When the user clicks on the edit button, we will show the expense popup form with the expense details filled in. The user can then edit the expense and save it.
+
+We will add a click event handler to the edit button in the `ExpenseCard` component.
```
Where did the expense card component come from? It's not in the tutorial?

Mridul2820 replied:
My mistake, It is ExpenseList

gewenyu99 reviewed (changes requested):
Hi friend, I think the content is really close.

I would take some time, follow your own tutorial from scratch and make sure everything works. 

I have some issues with the writing style that will be a lot of tweaking, too. But let's get the content down first.

We're really close :)

Mridul2820 commented:
> <img alt="Screenshot 2023-10-13 at 3 59 55 PM" width="860" src="https://user-images.githubusercontent.com/29069505/275031480-234c1233-1490-449c-8ea6-0c06e7d017ff.png">

@gewenyu99  I have resolved all the changes. Only this one is left. can you give me details on the page?

Mridul2820 commented:
Hi @gewenyu99 any update on this PR review?

gewenyu99 commented:
> > <img alt="Screenshot 2023-10-13 at 3 59 55 PM" width="860" src="https://user-images.githubusercontent.com/29069505/275031480-234c1233-1490-449c-8ea6-0c06e7d017ff.png">
> 
> @gewenyu99 I have resolved all the changes. Only this one is left. can you give me details on the page?

When I follow the tutorial and copy paste in the code, after registering I get this error :P

Mridul2820 commented:
> > > <img alt="Screenshot 2023-10-13 at 3 59 55 PM" width="860" src="https://user-images.githubusercontent.com/29069505/275031480-234c1233-1490-449c-8ea6-0c06e7d017ff.png">
> > 
> > 
> > @gewenyu99 I have resolved all the changes. Only this one is left. can you give me details on the page?
> 
> When I follow the tutorial and copy paste in the code, after registering I get this error :P

 I made some changes in the tutorial. Can you try it again?

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-6/+page.markdoc:
````diff
@@ -0,0 +1,193 @@
+
+Let's fetch the user expenses from the database and display them in a list. Open the `app/page.js` file and add the following code.
+
+```js
````
missing "use client"

gewenyu99 reviewed (changes requested):
I have some comments still, but a more serious problem is that the app is just not working for me.

I'm still facing the cookies issues like before, and the flow of the tutorial is just really difficult to follow. 

I think all the partial codes, long code files, skipped code with inconsistent format is too much to follow. Especially for the pagination buttons. I have no idea which code is meant to fo where.

There's also no conditional routing when the user isn't logged in. I'm taken to the expense page any, it seems, when when I'm not logged in.

I would encourage you to either refactor the code so the code snippets used in these examples are shorter or more concise, or make the app simpler.

Great work, I really liked most of the tutorial, it's just a little to complex to follow because of how it's presented.

Maybe take a look at other popular tutorials on Next.js and see if you can get some inspiration?

gewenyu99 commented:
I'm still getting the missing cookies error btw. Have you tried following your own tutorial carefully with a new repository?

Mridul2820 commented:
> I'm still getting the missing cookies error btw. Have you tried following your own tutorial carefully with a new repository?

Yes, I have tried to follow the tutorial on the new repo. And also made changes accordingly. Can you please try with a fresh installation?

Mridul2820 commented:
Hi @gewenyu99 I got your point

In fact, I created the project earlier in Next.js pages router which is much simpler and beginners friendly

But for Next.js app router you need to have at least the basics of Next is v13 and the is a minimum requirement it.

It is the most simplest app which covers all the basic Appwrite features

I can create a todo app but it also will have the same kind of stuff. 

The Next.js 13 app router is a complex I know. I covered the Appwrite topics in the tutorial

@TGlide also referred to use some advance topics of React and Next.js . So I completely changed some of the core concepts

gewenyu99 commented:
@Mridul2820 We shouldn't be making compromises there. I think we can do something about the way we present the flow of the apps.

Perhaps what we need is to have more checkpoints. Break up the implementation into checkpoints

-> At the end of auth should give a functioning app, it's a check point
-> At the end of create expenses give a functioning app,  it's a check point
-> Then add editing and deleting, gives a functioning app,  it's a check point

Think about what logic in this app is unnecessary.

gewenyu99 commented:
Is there also a way to break up the apps better in a way where we don't have to
- replace code in the middle of the file
- Have code examples that I need to search through to find what changed?

gewenyu99 commented:
I really like the app and the content, it just needs some work on presentation. Then we can work on the writing and presentation. 

Thanks!

Mridul2820 commented:
> @Mridul2820 We shouldn't be making compromises there. I think we can do something about the way we present the flow of the apps.
> 
> Perhaps what we need is to have more checkpoints. Break up the implementation into checkpoints
> 
> -> At the end of auth should give a functioning app, it's a check point -> At the end of create expenses give a functioning app, it's a check point -> Then add editing and deleting, gives a functioning app, it's a check point
> 
> Think about what logic in this app is unnecessary.

Can you specify how I can add the checkpoints?

Mridul2820 commented:
> Is there also a way to break up the apps better in a way where we don't have to
> 
> * replace code in the middle of the file
> * Have code examples that I need to search through to find what changed?

I have already separated the components for create, edit and delete in different files. But we have to join them together. right? 

Also how I can add file changed stuffs in the Appwrite markdoc files ?

gewenyu99 commented:
@Mridul2820 🤔 I'll try to take a look soon. Writing tutorials is tough 😄 But there's always a solution.

gewenyu99 commented:
Taking another look now

gewenyu99 commented:
@Mridul2820 Oh, another point @Mridul2820, we'll continue to work on this, if it's not ready by end of this month, we'll merge it into a different branch and give it a hacktoberfest-accepted label, so we can keep working on it. I might also get someone on our team to take over.

I can't say this enough, this is an amazing contribution, but immensely challenging. I really appreciate it!

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-4/+page.markdoc:
````diff
@@ -0,0 +1,217 @@
+};
+```
+
+After the session is created, we are creating a JWT token using `createJWT` method. We are storing the JWT token and user id in the cookie. Then we are redirecting the user to the home page.
````
Let's add something after this that points users to run the project and try these pages.

gewenyu99 commented:
Hold-on preview is not updating, let me update preview. Lemme run build from vercel :D

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-6/+page.markdoc:
````diff
@@ -0,0 +1,195 @@
+Let's fetch the user expenses from the database and display them in a list. Open the `app/page.js` file and add the following code.
+
+```js
+"use client";
````
For this change, the problem I still have is how confusing this example is.

The only part hidden is old code, but where did the rest of the code change? Can you break up this example?

Is there another way to express this change?

Maybe just show what's changed and use placeholder functions/empty functions and commented HTML placeholders in previous steps to make it more clear?

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-5/+page.markdoc:
```diff
@@ -0,0 +1,283 @@
+                className="input-text"
+                placeholder="Title"
+                name="amount"
+                type="number"
```
<img width="587" alt="Screenshot 2023-10-24 at 2 39 20 PM" src="https://github.com/appwrite/website/assets/29069505/f8e5c6d0-cecf-4117-a8ba-d6332fcced3b">
I think this is only taking ints

gewenyu99 commented on src/routes/docs/tutorials/nextjs/step-4/+page.markdoc:
```diff
@@ -0,0 +1,217 @@
+  try {
+    await account.create(ID.unique(), email, password, name);
+    await account.createEmailSession(email, password);
+    const jwt = await account.createJWT();
```
I think there's an issue here with JWTs expiring every 30 minutes 🤔

Mridul2820 replied:
it's expiring even early and I can't find a way to resolve this issue.

gewenyu99 replied:
@Mridul2820 We may have to wait for proper SSR support from our side 😬

When we accept this, we'll hold off on releasing it until we have proper SSR support.

gewenyu99 reviewed (commented):
Here's some technical problems I found upon further inspection of the code :)

Mridul2820 commented on src/routes/docs/tutorials/nextjs/step-6/+page.markdoc:
```diff
@@ -0,0 +1,195 @@
+  if (!userId || !jwt) {
+    return new NextResponse("Unauthorized", { status: 401 });
+  }
+  account.client.setJWT(jwt);
```
This is using the Appwrite JavaScript SDK 

I found the demo example at [demo-todo-with-nextjs](https://github.com/appwrite/demo-todo-with-nextjs)

Mridul2820 commented on src/routes/docs/tutorials/nextjs/step-4/+page.markdoc:
````diff
@@ -0,0 +1,217 @@
+
+```js
+// /src/app/(pages)/signup/page.js
+const signup = async (e) => {
````
> This needs to go in a very specific spot. in the exported component, �assuming I know nothing about the structure of a Next.js component, show me where this needs to go.

But this is a beginner-level tutorial for Appwrite and not for Next.js if I add all basic Next.js stuff the tutorial length will grow more. right? 

I have also followed The quick start guide of Appwrite's Next.js guide

gewenyu99 replied:
I think we can accomplish both, tutorials in general are directed at beginners.

More senior developers and experience developers tend to prefer to play around and experiment by themselves.

Mridul2820 commented on src/routes/docs/tutorials/nextjs/step-4/+page.markdoc:
````diff
@@ -0,0 +1,217 @@
+Now let's add the login function. We will be using Appwrite's `createEmailSession` method to create a new session. 
+
+```js
+const loginUser = async (e) => {
````
I have move the function to the same code section

gewenyu99 commented:
Here's an idea @Mridul2820 
https://github.com/appwrite/website/pull/275/files#diff-3780ba26d123aa2fd91e5a4906501aa4128fcaaa6f2c76c6e4f956de547cb84e

Take a look here, they explain the code in bits, but always include completed code at end of discussion, so it's easier to follow :)

gewenyu99 commented:
@Mridul2820 I think this is a good contribution, but it's blocked by a couple of things.

Some of the blockers are just that Appwrite doesn't play amazingly with SSR, and you've chosen to go with SSR for auth, so we can't merge this until SSR is released.

Some other blockers are just the complexity of this project. I think I'll take this and work with some Appwrite engineers to reduce the scope of this app.

If you'd like to continue working on this, feel free, but I believe this contribution is valuable enough for us as is.

Thank you so much for your patience and hard work!

Mridul2820 commented:
Cool @gewenyu99 

I will remove the SSR and add `useContext`

Also, I found the same kind of configuration for users and documents in the `sveltekit` tutorial earlier

gewenyu99 commented:
> Cool @gewenyu99
> 
> I will remove the SSR and add `useContext`
> 
> Also, I found the same kind of configuration for users and documents in the `sveltekit` tutorial earlier

Yeah, I'll accept your PR for hacktoberfest and keep it open for now.

We will add SSR to Appwrite properly, so we can take advantage of the amazing app you built.

Don't change your app for now, I think using SSR is correct way to use Next.js, the problem in on our end, and we'll work to address it.

I've marking this PR hacktoberfest-accepted so you'll get a valid contribution out of this.

Thanks so much!

gewenyu99 reviewed (approved):
For the purpose of this Hacktoberfest contribution, this is ready to be merged into a branch. I'll be taking over with the team to make some final touches when we have better SSR support and release this tutorial.

Mridul2820 commented:
Great, Thanks :)

gewenyu99 commented:
Hey there! There were a lot of big PRs during this Hacktoberfest, and we wanted to give everyone ample time to collaborate with our engineering team. If you were able to merge your PRs during October, amazing. If it’s still not merged, don’t worry about it either. Either way, we’ve got your Hacktoberfest swag minted and ready to ship. 

Please comment with your Discord username here so we can contact you about your shipping information to deliver your Hacktoberfest swag.

Mridul2820 commented:
> Discord username

@gewenyu99 Here is my discord username **mridul2820**

gewenyu99 commented:
Reaching out soon!
